File: SWEA/D3/2817. 부분 수열의 합/부분 수열의 합.java
import java.util.Scanner;

public class Solution {
	static int N, K;
	static int[] arr;
	static boolean[] check;
	static int cnt;

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int T = sc.nextInt();
		sc.nextLine();
		for (int tc = 1; tc <= T; tc++) {
			// 전체 입력 받기
			N = sc.nextInt();
			K = sc.nextInt();
			arr = new int[N];
			check = new boolean[N];
			cnt = 0;
			for (int i = 0; i < N; i++) {
				arr[i] = sc.nextInt();
			}
			sc.nextLine();
			powerset(0);
			System.out.println("#" + tc + " " + cnt);
		}
	}

	static void powerset(int idx) {
		// base case
		if (idx >= N) {
			int sum = 0;
			for (int i = 0; i<N; i++) {
				if (check[i]) {
					sum += arr[i];
				}
			}
			if (sum == K) {
				cnt++;
			}
			return;
		}

		// recursive case
		// check 에 넣는 부분
		check[idx] = true;
		powerset(idx + 1);

		check[idx] = false;
		powerset(idx + 1);
	}
}

File: SWEA/D3/2817. 부분 수열의 합/부분 수열의 합.py
def combination(idx):
    global N, K, cnt, check
    # base case
    if idx >= N:
        sum = 0
        for i in range(N):
            if check[i]:
                sum += arr[i]
        if sum == K:
            cnt += 1
        return

    # recursive
    check[idx] = True
    combination(idx+1)
    check[idx] = False
    combination(idx+1)


T = int(input())
for tc in range(1, T+1):
    N, K = map(int, input().split())
    arr = list(map(int, input().split()))
    cnt = 0
    check = [False] * N
    combination(0)
    print(f'#{tc} {cnt}')